### PR TITLE
Implement complex `logaddexp` and `logaddexp2`

### PR DESCRIPTION
**Pull request by wdphy16**
Implements @PhilipVinc 's proposal #6863.

Complex `logaddexp` and `logaddexp2` are not implemented in numpy yet, so I added custom tests `testLogaddexpComplex` and `testLogaddexp2Complex`, and provided baseline implementations there which are obviously correct but not so numerically stable. Eventually I'll upport them to numpy.

Another problem is that the imaginary part of complex log can differ by an arbitrary multiple of 2 pi. The convention of `np.log` is that the imaginary part lies in `[-pi, pi]` (which is inherited from the C language standard), but this convention is broken in scipy's complex `logsumexp`. We chose to follow this convention in complex `logaddexp` to reduce undefined behaviors.

There're a lot of edge cases involving `inf` and `nan` if we fully comply with the C language standard's `cexp` and `clog`, but I would not implement them with a lot of branches and performance loss. So there is no check for `inf` and `nan` in the code path for complex inputs, and I disabled the tests with `inf`. (By the way, we actually need to first fix such edge cases for `exp2`, `log2` and `log10` before fixing `logaddexp2`)

After this PR, all activation functions in `jax.nn` support complex numbers now.

### CODE REVIEW

**Comment by PhilipVinc** [comment]
wrt to the point raised in #6863, I wrote in the numpy mailing list about implementing complex-valued logaddexp, and it was pointed out that there's nothing against it (and actually, that tests should be added for complex-valued `scipy.logsumexp` because there are none).

@wdphy16 is working towards a PR to numpy but that will take a bit more time, so we are focusing on jax first.

File: tests/lax_numpy_test.py [hunk 1]
```diff
@@ -200,6 +200,18 @@ def op_record(name, nargs, dtypes, shapes, rng_factory, diff_modes,
...
+  if np.issubdtype(x1.dtype, np.complexfloating) or np.issubdtype(x2.dtype, np.complexfloating):
+    return np.log(np.exp(x1) + np.exp(x2))
+  else:
+    return np.logaddexp(x1, x2)
+
+def _logaddexp2_override(x1, x2):
+  if np.issubdtype(x1.dtype, np.complexfloating) or np.issubdtype(x2.dtype, np.complexfloating):
+    return np.log2(np.exp2(x1) + np.exp2(x2))
+  else:
+    return np.logaddexp2(x1, x2)
```
Comment (jakevdp) [inline tests/lax_numpy_test.py:213]: You can define these functions within tests of logaddexp and logaddexp2.

For and example of this, here is a test of functionality offered by `jnp.nonzero` that does not have an equivalent in `numpy`: https://github.com/google/jax/blob/cbfa539053d7cab3f07309504fb6bb5454dd91f2/tests/lax_numpy_test.py#L946-L952

Rather than trying to greate a general method to override numpy functions, you can just write a specific check for the behavior you want to check.

**Review by jakevdp — CHANGES_REQUESTED** [review]
Thanks for the contribution! All looks pretty good - my main comments involve simplifying the way this is tested - see below.

**Comment by wdphy16** [comment]
Thanks for your suggestions. Now I'm using custom tests for them.

File: tests/lax_numpy_test.py [hunk 2]
```diff
@@ -5379,6 +5379,53 @@ def testIssue2347(self):
...
+      for dtype in complex_dtypes))
+  def testLogaddexpComplex(self, shape, dtype):
+    rng = jtu.rand_some_nan(self.rng())
+    args_maker = lambda: (rng(shape, dtype), rng(shape, dtype))
+
+    @jtu.ignore_warning(category=RuntimeWarning, message="invalid value.*")
+    def np_fun(x1, x2):
+      return np.log(np.exp(x1) + np.exp(x2))
+
+    @jtu.ignore_warning(category=RuntimeWarning, message="invalid value.*")
```
Comment (jakevdp) [inline tests/lax_numpy_test.py:5396]: I don't think this decorator should be required on the jnp function

File: tests/lax_numpy_test.py [hunk 3]
```diff
@@ -5379,6 +5379,53 @@ def testIssue2347(self):
...
+    def jnp_fun(x1, x2):
+      y = jnp.logaddexp(x1, x2)
+      y = y.real + ((y.imag + np.pi) % (2 * np.pi) - np.pi) * 1j
```
Comment (jakevdp) [inline tests/lax_numpy_test.py:5399]: As written this would allow a future refactoring of `jnp.logaddexp` to return different outputs and still have the test pass. I think the logic for standardizing the complex branch of the output should go within the `logaddexp` function rather than its test. If possible, we should anticipate what branch `np.logaddexp` will likely use when it is implemented.

One idea: could we somehow ensure that `logaddexp(x1, x2)` will always choose the same branch as `log(x1 + x2)`?

File: tests/lax_numpy_test.py [hunk 4]
```diff
@@ -5379,6 +5379,53 @@ def testIssue2347(self):
...
+    @jtu.ignore_warning(category=RuntimeWarning, message="invalid value.*")
+    def np_fun(x1, x2):
+      return np.log2(np.exp2(x1) + np.exp2(x2))
+
+    @jtu.ignore_warning(category=RuntimeWarning, message="invalid value.*")
+    def jnp_fun(x1, x2):
+      y = jnp.logaddexp2(x1, x2)
+      pi_over_log2 = np.pi / np.log(2)
+      y = y.real + ((y.imag + pi_over_log2) % (2 * pi_over_log2) - pi_over_log2) * 1j
+      return y
```
Comment (jakevdp) [inline tests/lax_numpy_test.py:5424]: Same comments as above apply here as well.

**Review by jakevdp — CHANGES_REQUESTED** [review]
Looking great, thanks for reworking the tests!

One more substantive comment below, please let me know what you think

File: tests/lax_numpy_test.py (hunk 2 above)
Comment (wdphy16) [reply tests/lax_numpy_test.py:5396]: It's not for the jnp function, but `(y.imag + np.pi) % (2 * np.pi)` may raise the exception if `y.imag` contains `nan`.

File: tests/lax_numpy_test.py (hunk 3 above)
Comment (wdphy16) [reply tests/lax_numpy_test.py:5399]: My main concern is that adding a modulo operation into `jnp.logaddexp` will cause a little performance loss. Besides that, scipy's complex `logsumexp` doesn't seem to have such standardization, although it claims that it behaves the same as `np.log(np.sum(np.exp(a)))` in the document.

Actually, if we really want to fully reproduce the behavior of `np.log`, it requires more than a modulo, because the imaginary part of `np.log` can range in `[-pi, +pi]`, with both ends of the interval closed. For example, `np.log(np.pi*1j) = +pi`, and `np.log(-np.pi*1j) = -pi`.

Comment (jakevdp) [reply tests/lax_numpy_test.py:5399]: Sure, that makes sense if you want to keep the implementation as-is. But lets test the output in a way that validates the output of the function, rather than validating a post-processed version of the output.

Comment (PhilipVinc) [reply tests/lax_numpy_test.py:5399]: What are you suggesting we do for the test, then? 
It's rather hard to predict, given two outputs, what branch the output of `jnp.logaddexp`.

The sensible test should be to check that real parts match and imagnary parts differ by a multiple of 2*pi.
Should we avoid using `_CheckAgainstNumpy` and implement something like `_CheckAgainstNumpy2PiImagPhase`?

Comment (wdphy16) [reply tests/lax_numpy_test.py:5399]: If you agree that we will not standardize the branch of `jnp.logaddexp`, the valid test for complex log should not be 'real and image parts are the same', but 'real parts are the same, and image parts mod 2 pi are the same'. So I still think we need to first post-process the output by taking the mod, then send it to `_CheckAgainstNumpy`.

Maybe we should not use `_CheckAgainstNumpy` here, because its semantic would be 'real and image parts are the same'... but if we write our own check here, we would just copy a lot of code from it.

Comment (jakevdp) [reply tests/lax_numpy_test.py:5399]: > The branch may also change if we refactor the implementation in future.

This is what I'm hoping the test would catch. Over time users tend to write code that relies on such implementation details; it would be nice if the tests would let us know of such changes if and when they happen.

Comment (wdphy16) [reply tests/lax_numpy_test.py:5399]: Ok, then I will put the mod into `jnp.logaddexp`.

**Comment by wdphy16** [comment]
Now I implemented the modulus in `jnp.logaddexp` and ensures that the imaginary part of the output is in `[-pi, pi]`. I profiled some implementations of the modulus on my GPU and picked a faster one.

**Review by jakevdp — APPROVED** [review]
Looks great! Thanks for all the iteration on this.

**Comment by jakevdp** [comment]
CI failure seems unrelated... I'll try to track down the source of it.

**Comment by jakevdp** [comment]
Once #7166 is merged, you can rebase and then hopefully tests will pass

**Comment by jakevdp** [comment]
Actually, it looks like there are some failures in X64 mode, having to do with promotion between complex64 and complex128. You can run those tests locally with
```
$ JAX_ENABLE_X64=1 pytest tests/lax_numpy_test.py -k testOpGradLogaddexp
```

File: tests/lax_numpy_test.py [hunk 5]
```diff
@@ -5404,10 +5438,10 @@ def grad_test_spec(op, nargs, order, rng_factory, dtypes, name=None, tol=None):
                    tol={np.complex64: 2e-2}),
     grad_test_spec(jnp.logaddexp, nargs=2, order=1,
                    rng_factory=partial(jtu.rand_uniform, low=-0.9, high=0.9),
-                   dtypes=[np.float64], tol=1e-4),
-    grad_test_spec(jnp.logaddexp2, nargs=2, order=2,
+                   dtypes=[np.float64, np.complex128], tol=1e-4),
```
Comment (wdphy16) [inline tests/lax_numpy_test.py:5441]: It seems that changing `complex64` to `complex128` here solves the problem. Previously when the dtypes of the two operands are `pycomplex64` and `complex64`, even if I use things like `_promote_dtypes_inexact`, `pycomplex64` will somehow become `complex128`, while `complex64` is still `complex64`.

The specs above are using `complex64`, but they only have one operand. Should we really use `complex64` here?

Comment (jakevdp) [reply tests/lax_numpy_test.py:5441]: It would be best to track down where that unexpected promotion is happening

Comment (wdphy16) [reply tests/lax_numpy_test.py:5441]: I added some logging and found that when x64 is enabled, in `jax/test_util.py`, function `check_jvp`, line `v_out, t_out = f_jvp(args, tangent)`, `_promote_dtypes_inexact` will convert `complex128, complex64` to `complex128, complex128`; but in line `v_out_expected = f(*args)`, the same dtypes will be converted to `complex64, complex64`.

Maybe we need to understand things in `jax/_src/dtypes.py`, and how the x64 flag affects them.

Comment (jakevdp) [reply tests/lax_numpy_test.py:5441]: I suspect this may be related to weak types, owing to the change in processing the inputs.

Comment (jakevdp) [reply tests/lax_numpy_test.py:5441]: Ah, turns out this was a pre-existing issue. I went to the main branch and changed `float64` to `float32` in the same test, then ran it in X64 mode and got the same error message (here for a mismatch between float32 and float64)

Given that, I think your solution of using complex128 rather than complex64 here is probably the right one for this particular PR.

**Comment by jakevdp** [comment]
Last issue: on TPU, `testOpGradLogaddexp2` is failing with precision errors: it looks like the relative tolerance would need to be bumped up to around 0.05 for the tests to pass. You could probably do this by passing a tolerance that depends on `jtu.device_under_test() == 'tpu'`

**Comment by wdphy16** [comment]
I wrote the custom tests `testGradLogaddexpComplex` and `testGradLogaddexp2Complex`, and set the tolerance there.

Also, I looked again how to write custom tests, and added more combinations of shapes and dtypes to the tests of comparing against numpy.

It's true that the gradient of complex `logaddexp` can be more unstable than the real one:
```
d/dx1 log(exp(x1) + exp(x2)) = exp(x1) / (exp(x1) + exp(x2))
```
where `exp(x1)` can be a large positive number, and `exp(x2)` can be a large negative number, so the denominator can be small.

**Review by jakevdp — APPROVED** [review]

File: tests/lax_numpy_test.py [hunk 6]
```diff
@@ -5405,7 +5443,7 @@ def grad_test_spec(op, nargs, order, rng_factory, dtypes, name=None, tol=None):
     grad_test_spec(jnp.logaddexp, nargs=2, order=1,
                    rng_factory=partial(jtu.rand_uniform, low=-0.9, high=0.9),
                    dtypes=[np.float64], tol=1e-4),
-    grad_test_spec(jnp.logaddexp2, nargs=2, order=2,
+    grad_test_spec(jnp.logaddexp2, nargs=2, order=1,
```
Comment (jakevdp) [inline tests/lax_numpy_test.py:5446]: Is this change necessary?

File: tests/lax_numpy_test.py [hunk 7]
```diff
@@ -5379,6 +5379,44 @@ def testIssue2347(self):
...
+  def testLogaddexpComplex(self, shapes, dtypes):
+    @jtu.ignore_warning(category=RuntimeWarning, message="invalid value.*")
+    def np_op(x1, x2):
+      return np.log(np.exp(x1) + np.exp(x2))
+
+    rng = jtu.rand_some_nan(self.rng())
+    args_maker = lambda: tuple(rng(shape, dtype) for shape, dtype in zip(shapes, dtypes))
+    tol = {np.complex64: 1e-5, np.complex128: 5e-15}
+    self._CheckAgainstNumpy(_promote_like_jnp(np_op), jnp.logaddexp, args_maker, tol=tol)
+    self._CompileAndCheck(jnp.logaddexp, args_maker)
```
Comment (jakevdp) [inline tests/lax_numpy_test.py:5399]: pass `tol` to `_CompileAndCheck` as well (here and below). It's leading to failures on internal tests.

**Review by jakevdp — CHANGES_REQUESTED** [review]

File: tests/lax_numpy_test.py (hunk 6 above)
Comment (wdphy16) [reply tests/lax_numpy_test.py:5446]: I don't think so. Let me revert it.

File: tests/lax_numpy_test.py (hunk 7 above)
Comment (wdphy16) [reply tests/lax_numpy_test.py]: Fixed that.

Comment (wdphy16) [reply tests/lax_numpy_test.py]: Hi @jakevdp , I've made the requested changes and I think we can submit it to the internal tests again.

**Comment by jakevdp** [comment]
Sorry, just coming off a long weekend. Looks great - hopefully we can get this merged today!
